### ISSUE TITLE
Libpcap bpf string formatting

### ISSUE THREAD

**Issue by dimakhki** [the-tcpdump-group/libpcap#445]
I am getting some wierd errors when using libpcap pacp_compile Please let me know if these are known issues or I am making some bpf formating/ordering mistakes

For example:

1) if I pass (protocol_filter and vlan_filter), I see no packets: Like ((arp or udp port 234) and vlan 12) But if I swap exp as (vlan_filter and protocol_filter), then packet filtering works: Like ((vlan 12) and (arp or udp port 234))

2) If I include vlan with "not" protocols, I get packets with protocol filter applied: eg: (((vlan 20)) and (not (udp port 3800)) and ((not (tcp)) and (not (udp port 53)) and (not (icmp)) and (not (ip6)) and (not (udp port 5353))))

If I dont include vlan with protocols, I get all packets including tcp, icmp, ip6 etc when bpf is: (not (udp port 3800)) and ((not (tcp)) and (not (udp port 53)) and (not (icmp)) and (not (ip6)) and (not (udp port 5353))))

3)(((not vlan)) and (not (udp port 3800)) and ((not (tcp)) and (not (udp port 53)) and (not (icmp)) and (not (ip6)) and (not (udp port 5353)))) get all packets including ip6, tcp, etc

(((not vlan)) and (not (udp port 3800)) and (((tcp)) or ((udp port 53)) or ((icmp)) or ((ip6)) or ((udp port 5353)))) dont get even single packet

**Comment by msekletar** [comment]
Practically the same issue was brought up numerous times (dig through the log here or on tcpdump-workers list).

tldr; if you are using vlan, mpls or ip6 in your filters you really need to know what are you doing. Yes, order in which these primitives appear in filter expression matters. Things will not work properly if you don't understand how are they handled internally. Sadly, these kind of gotchas are very difficult if not impossible to fix with current BPF compiler design.

**Comment by 1apodaca3** [comment]
Delete unwanted data

**Comment by 1apodaca3** [comment]
Don't want on phone

**Comment by 1apodaca3** [comment]
Unwanted

**Comment by dimakhki** [comment]
Srry Pushkar,
Though ur expression((not <filter> and not vlan) or (vlan and not
<filter>)) for case-3 solved "untagged ~P packets" problem now its not
working for "tagged P packets"

Regards
Sameer

**Comment by infrastation** [comment]
Looks like Michal had addressed this issue a long time ago. For reference, the [man page](https://www.tcpdump.org/manpages/pcap-filter.7.html) says:
> Note that the first vlan keyword encountered in an expression  changes  the  decoding  offsets for the remainder of the expression on the assumption that the packet is a VLAN packet.

With that and several VLAN-related bugfixes made in subsequent libpcap releases, let's close this issue.
